### ISSUE TITLE
[Improvement] Do not hardcode the "default" machine name, use DOCKER_MACHINE_NAME and default as fallback

### ISSUE THREAD

**Issue by Yivan** [docker/kitematic#3521]
First, thanks for this great software.

I would like to make a proposition of improvement.

Actually the machine name is hardcoded to "`default`".
There has been a patch (https://github.com/docker/kitematic/pull/992), but it don't work anymore (i tried it without any success) and found no more reference of it in the actual source code.

Actually the `default `machine name seems to be hardcoded:
https://github.com/docker/kitematic/blob/master/src/utils/DockerMachineUtil.js#L24

This is problematic because we don't use the name `default` in virtualbox (because we have others VM managing LXD, some Debians, etc.), and `default` mean nothing in the list of VM, it is not a clean name for developpers to easily spot the VM in charge of docker toolbox and can be prone error when cleaning VM. (actually we name our VM `DockerToolbox`)
So a solution of forcing naming docker VM "default" is not possible for us.

First question please: is it a way to change this `default` name ?

Second question: If not can an improvement be made on this ?
An easy solution would be to look for the environement variable DOCKER_MACHINE_NAME and if it exist take this one (this environment variable is used by docker-machine itself to determine the active machine). If this environement variable don't exist or is not set use `default`. This condition could be implemented here:
https://github.com/docker/kitematic/blob/master/src/utils/DockerMachineUtil.js#L24

Actually we can no more use Kinematic and switched to Portainer which handle well this. We would really like to able to use it again.

Thanks!

**Comment by Yivan** [comment]
The patch could simply be:
In `src/utils/DockerMachineUtil.js`, line 24, replace:
`return 'default';`
by
`return (process.env.DOCKER_MACHINE_NAME ? process.env.DOCKER_MACHINE_NAME : 'default');`

It will allow all user which don't use default as VM name to be able to use Kinematic again.

**Comment by Yivan** [comment]
Up please, someone here ? Could the patch be implemented for next release ?

**Comment by FrenchBen** [comment]
The behavior would need to first change in machine - `default` has always been the go-to for deployments and thus somewhat "reserved" for the Boot2docker provisioned VM. 

What happens when the env var is set, but the machine doesn't exist?

**Comment by Yivan** [comment]
Thanks @FrenchBen for your post.

Machinge get default machine name as `default` name, but it can be used with other names (`docker-machine [command] [machine name]`). Machine work great with other names.

Problem with Kinematic is it won't work at all with other names. `default` name is hardcoded on it and there is no way to indicate it to use another machine name. When Kinematic start if it don't find the default name machine, it even attempt to create it... : (

To answer to your question, i think if env var is set and machine is not found, it can do any of the following (by order of best preference):
- send an error message (i don't think it is the purpose of Kinematic to create the VM... so for me it is the best, and it give the developper the possibility to create manually the machine)
- create the machine with this name (like it does actually for default)
- create a default machine and use it

But the real problem in this issue is Kinematic cannot be used with machine with other name than `default`, and it limit its usablility. Docker Toolbox do not force to use default. You can create any new VM name you want with docker-machine.

And the patch is so easy to implement, and will bring so more to Kinematic, that actually i don't see any raison to not implement it. We don't talk about big code change here.

**Comment by FrenchBen** [comment]
Here's part of the complexity, say you're doing a lot of `docker-machine` things and to make your life easier, you set the ENV var to be machine.
In order for the ENV var to be "seen" by Kitematic, Kitematic would need to be started from the CLI (not very convenient for most users)
This VM isn't a boot2docker image, but it's what you use for X project.
You then decide to open Kitematic from CLI, as you've done before, with the changes you've mentioned above. 
It wont connect. You forgot that you had the ENV, and think Kitematic is being silly not finding your `default` vm and click the "Delete and try setup again". You've now blown away your X project machine and may be mad at Kitematic (even though you told it to do it)

TL;DR - it's easier to tell people that `default` is the Kitematic preferred VM, and the one that should always be used for boot2docker things, rather than providing complex options as to what VM/Driver combo should be used.

**Comment by Yivan** [comment]
I undestand better thanks for the explaination.
I was not aware that:
- Kinematic cannot read Windows system/user environment variables
- Kinematic has the job to manage the VM (create, etc.) (i thought Kinematic was only to view container in a docker instance)
- Kinematic has a "delete VM" policie if it don't manage to connect to VM

I am just not agree on you last sentence "default is the Kitematic preferred VM, and the one that should always be used for boot2docker things". It is more right to say Kinematic actually can only be used with default, this is not a preferred VM as there is no choice (and this is purpose of this issue which could be resolved i hope). For boot2docker, it has never been mandatory to use default, using another name works very great, just Kinematic don't handle it. In fact developpers could even have several boot2docker VM, this is the puprose of docker-machine, be able to create/manage multipe machine and so multiple boot2docker. Forcing it to only one VM called default is reducting. For instance, what about swarm in this case ? Or testing local cluster environment, etc...

I think if Kinematic cannot read windows environment the old patch could be reincluded (kinematic read a config file). To prevent machine deletion, if the VM kinematic try to start is not default, it could just report "Cannot connect" and that's all. It should not have an option to destroy and recreate... It is to the user to do it so VM cannot be deleted accidentaly. And we can consider that an user using not the default name is more advanced than a user staying on default so he know how to handle the problem.
What do you think of this approach?

p.s: does your deletion exemple apply to default too ? If i create a "default" VM for some project (imagine a CentOS VM) and start Kinematic it can destroy it ? If yes, deletion problem is not particularly tied to not use the `default` name for the VM in this case but more an other global issue i think.

**Comment by FrenchBen** [comment]
@Yivan there is a "choice", as the software is OSS and thus can be modified by you to do what you want. 
The origin of the tools were as such and thus have remained. I understand your need, but also see the simplicity of keeping a Docker tool as the default target. 
Even though it can now do much more, there's quite a lot that's ok to have as the default. 

Kitematic was suppose to prevent the hard-entry into the Docker CLI by simplifying the creation and connection to the VM. If the default was taken by something else, than Kitematic would not be able to connect to it. It will then give you the option to destroy it in order to start fresh, but will not do it without you clicking the button.

**Comment by Yivan** [comment]
I understand your reflexion and need to keep it simple. Thanks for taking time talking about this issue.
I hope somedays we could use kinematic with different host/VM name (like portainer does for instance).

**Comment by c33s** [comment]
i would vote for a reopen. having many virtualbox machines up and running on my system and  a machine which is named `default` inbetween them is really really unhandy.  quite a bad DX

i changed the env var to `docker` and assumed that at least after a host restart kitematic will be able to read my new set environment variable.

as docker toolbox introduced a env variable that is named `DOCKER_MACHINE_NAME` i would thinkg kitenmatic should obey this naming/config.
normally people would change it and if you provide some info in the readme then people who change the name, will know the risks.

just my 2 ct
